File: Algorithmic Toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/gcd.cpp
#include <iostream>

int gcd_naive(int a, int b) {
  int current_gcd = 1;
  for (int d = 2; d <= a && d <= b; d++) {
    if (a % d == 0 && b % d == 0) {
      if (d > current_gcd) {
        current_gcd = d;
      }
    }
  }
  return current_gcd;
}
int gcd(int a, int b) {
  if(b == 0) {
    return a;
  }

  int r = a % b;
  return gcd(b,r);
  }

int main() {
  int a, b;
  std::cin >> a >> b;
  std::cout << gcd(a, b) << std::endl;
  return 0;
}

File: Algorithmic Toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/gcd.py
# Uses python3
import sys


def gcd_naive(a, b):
    current_gcd = 1
    for d in range(2, min(a, b) + 1):
        if a % d == 0 and b % d == 0:
            if d > current_gcd:
                current_gcd = d

    return current_gcd


def efficient(a, b):
    if(b == 0):
        return a
    else:
        return efficient(b, a % b)


if __name__ == "__main__":
    input = sys.stdin.read()
    a, b = map(int, input.split())
    print(efficient(a, b))
